inferno/callbacks.py

```diff
@@ -186,24 +249,32 @@ def initialize(self):
                     "Wrong sign {}, expected one of {}."
                     "".format(sign, ", ".join(map(str, signs_allowed))))
 
+    def initialize(self):
         items = chain(self.default_key_signs.items(), self.key_signs.items())
         self.best_losses_ = {key: -1 * sign * np.inf for key, sign in items}
         return self
 
+    def _check_keys_missing(self, history):
+        check_keys = [k for k in self.key_signs if not self._is_optional(k)]
+        sl = np.s_[-1, check_keys]
+        check_history_slice(history, sl)
+
     def _yield_key_sign_loss(self, history):
-        for key, sign in self.default_key_signs.items():
-            try:
-                loss = history[-1, key]
-                yield key, sign, loss
-            except KeyError:
-                pass
-        for key, sign in self.key_signs.items():
-            loss = history[-1, key]
-            yield key, sign, loss
+        key_signs = chain(
+            self.default_key_signs.items(), self.key_signs.items())
+        for key_loss, sign in key_signs:
+            if self._is_optional(key_loss):
+                try:
+                    loss = history[-1, key_loss]
+                    yield key_loss, sign, loss
+                except KeyError:
+                    pass
+                continue
+            loss = history[-1, key_loss]
+            yield key_loss, sign, loss
```

Comment:
this can be simplified similar to the block above (i.e. raise on error, try always)